examples/name-service/programs/name-service/src/lib.rs

```diff
@@ -0,0 +1,249 @@
+            cpi_context,
+        };
+
+        verify(ctx, &inputs, &[&signer_seeds])?;
+
+        Ok(())
+    }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, BorshDeserialize, BorshSerialize)]
+pub enum RData {
+    A(Ipv4Addr),
+    AAAA(Ipv6Addr),
+    CName(String),
+}
+
+impl anchor_lang::IdlBuild for RData {}
+
+#[derive(Debug, BorshDeserialize, BorshSerialize, LightDiscriminator)]
+pub struct NameRecord {
+    pub owner: Pubkey,
+    pub name: String,
+    pub rdata: RData,
+}
+
+#[error_code]
+pub enum CustomError {
+    #[msg("No authority to perform this action")]
+    Unauthorized,
+    #[msg("Record account has no data")]
+    NoData,
+}
+
+#[light_accounts]
+#[derive(Accounts, LightTraits)]
+pub struct NameService<'info> {
+    #[account(mut)]
+    #[fee_payer]
+    pub signer: Signer<'info>,
+    #[self_program]
+    pub self_program: Program<'info, crate::program::NameService>,
+    /// CHECK: Checked in light-system-program.
+    #[authority]
+    pub cpi_signer: AccountInfo<'info>,
+}
+
+impl light_hasher::DataHasher for NameRecord {
+    fn hash<H: Hasher>(&self) -> std::result::Result<[u8; 32], HasherError> {
+        let owner = hash_to_bn254_field_size_be(self.owner.to_bytes().as_slice())
+            .unwrap()
+            .0;
+        H::hashv(&[&owner, self.name.as_bytes()])
+    }
+}
+
+fn signer_check(
+    ctx: &Context<'_, '_, '_, '_, NameService<'_>>,
+    compressed_account: &PackedCompressedAccountWithMerkleContext,
+) -> Result<()> {
+    let record = NameRecord::deserialize(
+        &mut compressed_account
+            .compressed_account
+            .data
+            .as_ref()
+            .ok_or(CustomError::NoData)?
+            .data
+            .as_slice(),
+    )?;
+    if ctx.accounts.signer.key() == record.owner {
+        Ok(())
+    } else {
+        err!(CustomError::Unauthorized)
+    }
+}
+
+fn create_compressed_account(
+    ctx: &Context<'_, '_, '_, '_, NameService<'_>>,
+    record: &NameRecord,
+    new_address_params: &NewAddressParamsPacked,
+) -> Result<OutputCompressedAccountWithPackedContext> {
+    let data = record.try_to_vec()?;
+    let data_hash = record.hash::<Poseidon>().map_err(ProgramError::from)?;
+    let compressed_account_data = CompressedAccountData {
+        discriminator: NameRecord::discriminator(),
+        data,
+        data_hash,
+    };
+    let address = derive_address(
+        &ctx.remaining_accounts[new_address_params.address_merkle_tree_account_index as usize]
+            .key(),
+        &new_address_params.seed,
+    )
+    .map_err(|_| ProgramError::InvalidArgument)?;
+    let compressed_account = CompressedAccount {
+        owner: crate::ID,
+        lamports: 0,
+        address: Some(address),
+        data: Some(compressed_account_data),
+    };
+
+    Ok(OutputCompressedAccountWithPackedContext {
+        compressed_account,
+        merkle_tree_index: 0,
+    })
+}
+
+fn compressed_account_with_address(
```

Comment:
can we rename since this only creates output compressed accounts?
```suggestion
fn compressed_output_account_with_address(
```

Reply:
Done.